### PR TITLE
feat: add support for implicit return

### PR DESCRIPTION
**Pull request by amchiclet**
One more attempt, let's goooo!

Partially helps with #8408 (but doesn't close it).

In languages such as Julia, Ruby, and Rust, an expression that is last executed in a function is implicitly that function's return value.

This PR supports matching these implicit returns.

For example, the pattern `return 0` can match functions such as
  ```julia
  f() = 0
  ```
and
  ```julia
  function f()
    0
  end
  ```
and more.

This feature can be disabled through the `implicit_return` rule option.

Most of the changes are similar to https://github.com/returntocorp/semgrep/pull/8826 with some notable changes:

* Added `CFG_build.cfg_of_fdef` which is wraps the logic of creating a CFG, mark nodes, and re-creating the CFG for languages that support implicit return. There are 4 call sites for this function now. I thought it made sense to include all the logic within the function and not the call sites.

* Updated `AST_to_IL.call_generic`. It was calling `AST_generic.e` every time to create a new expression node, so any changes to is_implicit_return is not persisted across CFG builds for call nodes that were created by `call_generic`.

* Removed an unneeded parameter from `AST_to_IL.cond_with_pre_stmts`.

* One taint_if maturity test in Elixir is failing after this change. However, the test was previously passing for the wrong reasons. I looked at the CFG in HEAD, and it was already wrong, but the tests just happened to pass. So I've marked the missing match as a TODO for now, and it will fixed later, separately.

Tested with `make core-test` and added new tests for Julia, Ruby, and Rust.

As for Elixir, the language support is still experimental. I'll add more tests later as it matures.

### CODE REVIEW

**Comment by amchiclet** [comment]
Talked to @ihji offline, I'm going to split this single commit into 2 commits:
* reverting back from https://github.com/returntocorp/semgrep/pull/8955 so it's the same state as what was reviewed before (https://github.com/returntocorp/semgrep/pull/8826)
* the fix on top of that state

The reviewer then can choose to only see what's new by comparing the second commit to the first commit. (Still same PR here.)

I'll comment here again once the commits are split.

**Comment by amchiclet** [comment]
I've split it into 2 commits.

Now it should be easier for the reviewer to see the "revert-revert" changes vs the fix on top of it.

File: changelog.d/gh-8408.added [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+can now match 0 in
+
+```julia
+function f()
+  0
+end
+```
+
+This matching is enabled by default and can be disabled with the rule option `implicit_return`.
```
Comment (IagoAbal) [inline changelog.d/gh-8408.added:17]: You checked this box:
>  Documentation is up-to-date

but have you opened a PR to semgrep-docs or added a ticket to the Documentation project? (Just checking.)

Comment (amchiclet) [reply changelog.d/gh-8408.added:17]: Oh, my bad! Thanks for checking.

I misunderstood "documentation". I thought it was about documenting the code, e.g. adding comments explaining things, etc, but now that I look at it again, it seems to fall into the "code is evident to future readers" bullet point.

I will add a ticket to track it + make changes to semgrep-docs.

File: src/analyzing/AST_to_IL.ml [hunk 2]
```diff
@@ -865,8 +865,7 @@ and expr_lazy_op env op tok arg0 args eorig =
   in
   mk_e (Operator ((op, tok), arg0' :: args')) eorig
 
-and call_generic env ?(void = false) tok e args =
-  let eorig = SameAs (G.Call (e, args) |> G.e) in
```
Comment (IagoAbal) [inline src/analyzing/AST_to_IL.ml:877]: Do we have a test that would fail if we ever undo this change by mistake?

File: src/analyzing/AST_to_IL.ml [hunk 3]
```diff
@@ -1133,10 +1137,10 @@ and stmt_expr_with_pre_stmts env st =
   with_pre_stmts env (fun env -> stmt_expr env st)
 
 (* alt: could use H.cond_to_expr and reuse expr_with_pre_stmts *)
-and cond_with_pre_stmts env ?void cond =
+and cond_with_pre_stmts env cond =
```
Comment (IagoAbal) [inline src/analyzing/AST_to_IL.ml:1148]: I think it makes sense to remove `?void` from here, but is this a clean up or a change needed for implicit-return ?

File: src/analyzing/AST_to_IL.ml [hunk 4]
```diff
@@ -1208,10 +1212,10 @@ and type_ env (ty : G.type_) : type_ =
  * use 'expr_with_pre_stmts' or other '*_pre_stmts*' functions. Just so that
  * we don't forget about 'env.stmts'! *)
 
-(* TODO: What other languages have no fallthrough? *)
 and no_switch_fallthrough : Lang.t -> bool = function
   | Go
-  | Ruby ->
+  | Ruby
+  | Rust ->
```
Comment (IagoAbal) [inline src/analyzing/AST_to_IL.ml:1226]: Is this related to implicit-return ?

File: src/analyzing/Constant_propagation.ml [hunk 5]
```diff
@@ -814,14 +814,18 @@ let propagate_dataflow lang ast =
       let xs =
         AST_to_IL.stmt lang (G.Block (Tok.unsafe_fake_bracket ast) |> G.s)
       in
+      (* No implicit return analysis needed for top-level scope,
```
Comment (IagoAbal) [inline src/analyzing/Constant_propagation.ml]: Why do we have this comment about implicit-return in a module that only concerns constant propagation?

File: src/core_cli/Core_actions.ml [hunk 6]
```diff
@@ -85,6 +85,14 @@ let dump_il file =
     pr2 s;
     pr2 "==>";
 
+    (* Creating a CFG and throwing it away here here so the implicit return
```
Comment (IagoAbal) [inline src/core_cli/Core_actions.ml]: "here here"

File: src/analyzing/tests/Test_analyze_generic.ml [hunk 7]
```diff
@@ -89,8 +88,10 @@ let test_dfg_svalue ~parse_program file =
       inherit [_] AST_generic.iter_no_id_info
 
       method! visit_function_definition _ def =
-        let inputs, xs = AST_to_IL.function_definition lang def in
-        let flow = CFG_build.cfg_of_stmts xs in
+        let inputs, flow =
+          match CFG_build.cfg_of_fdef lang def with
+          | { fparams; fcfg } -> (fparams, fcfg)
```
Comment (IagoAbal) [inline src/analyzing/tests/Test_analyze_generic.ml]: You can write this as `let {fparams = inputs; fcfg = flow} = CFG_build.cfg_of_fdef lang def in`.

File: src/analyzing/tests/Test_analyze_generic.ml [hunk 8]
```diff
@@ -64,9 +64,8 @@ let test_cfg_il ~parse_program file =
   Naming_AST.resolve lang ast;
   Visit_function_defs.visit
     (fun _ fdef ->
-      let _, xs = AST_to_IL.function_definition lang fdef in
-      let cfg = CFG_build.cfg_of_stmts xs in
-      Display_IL.display_cfg cfg)
+      match CFG_build.cfg_of_fdef lang fdef with
+      | { fparams = _; fcfg } -> Display_IL.display_cfg fcfg)
```
Comment (IagoAbal) [inline src/analyzing/tests/Test_analyze_generic.ml]: No need for a `match`, IMO a `let` is more clear when there is _just one_ case to match. But here I'm just nit picking.

File: src/analyzing/Constant_propagation.ml [hunk 9]
```diff
@@ -814,14 +814,18 @@ let propagate_dataflow lang ast =
...
+       *)
       let flow = CFG_build.cfg_of_stmts xs in
       propagate_dataflow_one_function lang [] flow
   | _ ->
       ast
       |> Visit_function_defs.visit (fun _ent fdef ->
-             let inputs, xs = AST_to_IL.function_definition lang fdef in
-             let flow = CFG_build.cfg_of_stmts xs in
-             propagate_dataflow_one_function lang inputs flow);
+             match CFG_build.cfg_of_fdef lang fdef with
```
Comment (IagoAbal) [inline src/analyzing/Constant_propagation.ml]: `let { ... } = ... in` (nit pick)

File: src/analyzing/CFG_build.mli [hunk 10]
```diff
@@ -1 +1,17 @@
+(* Compute the control flow graph of a sequence of statements.
+ * This sequence of statements doesn't necessarily have to be
+ * a function body, so no implicit return analysis is done here.
+ *)
 val cfg_of_stmts : IL.stmt list -> IL.cfg
```
Comment (IagoAbal) [inline src/analyzing/CFG_build.mli:8]: Perhaps call this `cfg_of_top_level_stmts` ?

File: src/analyzing/Constant_propagation.ml (hunk 5 above)
Comment (IagoAbal) [reply src/analyzing/Constant_propagation.ml]: OK I see now, I think it's fine to have these comments, but I would put all the info in `CFG_build.mli` and then here just have a very short comment, one can always go to `CFG_build.mli` to know more.

File: src/analyzing/AST_to_IL.ml [hunk 11]
```diff
@@ -1236,55 +1240,82 @@ and mk_switch_break_label env tok =
...
+   *)
+  let ss, e = expr_with_pre_stmts ~void:false env eorig in
+  let ret = mk_s (Return (tok, e)) in
+  ss @ [ ret ]
+
+and expr_stmt env (eorig : G.expr) tok : IL.stmt list =
+  (* optimize? pass context to expr when no need for return value? *)
+  let ss, e = expr_with_pre_stmts ~void:true env eorig in
+
+  (* ExprStmts don't care about results of an expression, so it is rightly
```
Comment (IagoAbal) [inline src/analyzing/AST_to_IL.ml]: Maybe we should just not create an `mk_aux_var` when something returns `()` ?

File: src/analyzing/AST_to_IL.ml [hunk 12]
```diff
@@ -1236,55 +1240,82 @@ and mk_switch_break_label env tok =
...
+  | { e = G.Call _; _ } -> ()
+  | _else_ -> mk_aux_var env tok e |> ignore);
+
+  let ss' = pop_stmts env in
+  match ss @ ss' with
+  | [] ->
+      (* This case may happen when we have a function like
+       *
+       *   function some_function(some_var) {
+       *     some_var
```
Comment (IagoAbal) [inline src/analyzing/AST_to_IL.ml:1290]: In this case, would not have the `mk_aux_var` above have created an assignment ?

File: src/analyzing/AST_to_IL.ml (hunk 2 above)
Comment (amchiclet) [reply src/analyzing/AST_to_IL.ml:877]: Good catch. I remember I had a test, but turns out I don't.

I just added a test for this now.

File: src/analyzing/AST_to_IL.ml (hunk 3 above)
Comment (amchiclet) [reply src/analyzing/AST_to_IL.ml:1148]: It's a clean up that I found when I was trying to make sense of the use of `?void` in implicit returns. Not directly related.

File: src/analyzing/AST_to_IL.ml (hunk 4 above)
Comment (amchiclet) [reply src/analyzing/AST_to_IL.ml:1226]: I don't remember what actually brought my attention to this function. Not related to implicit return.

File: src/analyzing/AST_to_IL.ml (hunk 12 above)
Comment (amchiclet) [reply src/analyzing/AST_to_IL.ml:1290]: Calling `mk_aux_var` doesn't guarantee an assignment. When the expression passed is a Fetch, there's no assignment created. Previously I think I might have proposed some changes to `mk_aux_var`, but that was quite risky because `mk_aux_var` is used in many places.

File: src/analyzing/AST_to_IL.ml (hunk 11 above)
Comment (amchiclet) [reply src/analyzing/AST_to_IL.ml]: Interesting. I like that idea. Let me try it.

File: src/analyzing/Constant_propagation.ml (hunk 5 above)
Comment (amchiclet) [reply src/analyzing/Constant_propagation.ml]: Sounds good! I'll make the change.

File: src/analyzing/tests/Test_analyze_generic.ml (hunk 8 above)
Comment (amchiclet) [reply src/analyzing/tests/Test_analyze_generic.ml]: I never knew about this pattern, so this nit picking is very useful!

File: src/analyzing/CFG_build.mli (hunk 10 above)
Comment (amchiclet) [reply src/analyzing/CFG_build.mli:8]: Changing the name of this function will break pro while waiting for PR reviews and syncing between versions of the oss submodule. If possible, I'd like to make sure this PR integrates well with pro first, then I can rename this function later.

File: src/analyzing/Constant_propagation.ml (hunk 9 above)
Comment (amchiclet) [reply src/analyzing/Constant_propagation.ml]: Done.

File: src/analyzing/tests/Test_analyze_generic.ml (hunk 7 above)
Comment (amchiclet) [reply src/analyzing/tests/Test_analyze_generic.ml]: Done.

File: src/core_cli/Core_actions.ml (hunk 6 above)
Comment (amchiclet) [reply src/core_cli/Core_actions.ml]: Fixed typo.

File: src/analyzing/AST_to_IL.ml (hunk 11 above)
Comment (amchiclet) [reply src/analyzing/AST_to_IL.ml]: It works!

**Comment by amchiclet** [comment]
Done addressing the feedback. Thanks.

**Review by IagoAbal — APPROVED** [review]

File: src/analyzing/Implicit_return.ml [hunk 13]
```diff
@@ -36,39 +36,41 @@ open AST_generic
...
-   *)
-  mark_first_instr_ancestor cfg cfg.exit
-
 (*****************************************************************************)
 (* Entry point *)
 (*****************************************************************************)
 
 let lang_supports_implicit_return (lang : Lang.t) =
   match lang with
+  | Elixir
```
Comment (ihji) [inline src/analyzing/Implicit_return.ml:65]: Should we also support Scala?

**Review by ihji — APPROVED** [review]

File: src/analyzing/Implicit_return.ml (hunk 13 above)
Comment (amchiclet) [reply src/analyzing/Implicit_return.ml:65]: Sure. Added scala tests. Lots of them are passing, but some aren't, e.g. the IL has FIXME nodes, so for those test cases, I made them TODOs, and I'll create a ticket to track making all test cases work for scala.

**Review by aryx — APPROVED** [review]
Let's merge this!
